core/src/main/java/io/questdb/cutlass/line/tcp/DelegatingTlsChannel.java

```diff
@@ -0,0 +1,367 @@
+import java.lang.reflect.Field;
+import java.nio.Buffer;
+import java.nio.ByteBuffer;
+import java.security.KeyManagementException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.cert.CertificateException;
+
+public final class DelegatingTlsChannel implements LineChannel {
+    private static final int INITIAL_BUFFER_CAPACITY = 64 * 1024;
+    private static final long ADDRESS_FIELD_OFFSET;
+    private static final long LIMIT_FIELD_OFFSET;
+    private static final long CAPACITY_FIELD_OFFSET;
+
+    private static final int INITIAL_STATE = 0;
+    private static final int AFTER_HANDSHAKE = 1;
+    private static final int CLOSING = 2;
+    private static final int CLOSED = 3;
+
+
+    private final LineChannel delegate;
+    private final SSLEngine sslEngine;
+
+    private final ByteBuffer wrapInputBuffer;
+    private ByteBuffer wrapOutputBuffer;
+    private ByteBuffer unwrapInputBuffer;
+    private ByteBuffer unwrapOutputBuffer;
+    private final ByteBuffer dummyBuffer;
+
+    private long wrapOutputBufferPtr;
+    private long unwrapInputBufferPtr;
+    private long unwrapOutputBufferPtr;
+
+    private int state = INITIAL_STATE;
+
+    static {
+        Field addressField;
+        Field limitField;
+        Field capacityField;
+        try {
+            addressField = Buffer.class.getDeclaredField("address");
+            limitField = Buffer.class.getDeclaredField("limit");
+            capacityField = Buffer.class.getDeclaredField("capacity");
+        } catch (NoSuchFieldException e) {
+            // todo: consider a fallback strategy - we could keep copying buffers
+            throw new LineSenderException("unexpected buffer");
+        }
+        ADDRESS_FIELD_OFFSET = Unsafe.getUnsafe().objectFieldOffset(addressField);
+        LIMIT_FIELD_OFFSET = Unsafe.getUnsafe().objectFieldOffset(limitField);
+        CAPACITY_FIELD_OFFSET = Unsafe.getUnsafe().objectFieldOffset(capacityField);
+    }
+
+    public DelegatingTlsChannel(LineChannel delegate, String trustStorePath, char[] password) {
+        this.delegate = delegate;
+        this.sslEngine = createSslEngine(trustStorePath, password);
+
+        // wrapInputBuffer is just a placeholder, we set the internal address, capacity and limit in send()
+        this.wrapInputBuffer = ByteBuffer.allocateDirect(0);
+
+        // allows to override in tests, but we dont neccessary want to expose this to users.
+        int initialCapacity = Integer.getInteger("questdb.experimental.tls.buffersize", INITIAL_BUFFER_CAPACITY);
+
+        this.wrapOutputBuffer = ByteBuffer.allocateDirect(initialCapacity);
+        this.unwrapInputBuffer = ByteBuffer.allocateDirect(initialCapacity);
+        this.unwrapOutputBuffer = ByteBuffer.allocateDirect(initialCapacity);
+        this.wrapOutputBufferPtr = Unsafe.getUnsafe().getLong(wrapOutputBuffer, ADDRESS_FIELD_OFFSET);
```

Comment:
Can we perhaps inject pointer into these buffers too, so that we able to free memory when this channel is closed ? We can also track leaks

Reply:
that's a good idea, will do that!